File: fhirstarter/src/main/java/at/spengergasse/fhirstarter/entity/Encounter.java
package at.spengergasse.fhirstarter.entity;

import at.spengergasse.fhirstarter.model.*;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "en_encounter")
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@SuperBuilder
public class Encounter extends DomainResource {

    public enum Statuscode {planned, arrived, triaged, @JsonProperty("in-progress") in_progress, onleave, finished, cancelled};


    @OneToMany(cascade = CascadeType.ALL)
    @JoinColumn(name = "en_id_identifier")
    private List<Identifier> identifiers = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    @Column(name = "en_status", nullable = false) //hier bei @Column für DB
    @NotNull //hier für den Controller (http Requests
    private Statuscode status;

    @OneToMany(cascade = CascadeType.ALL)
    @JoinColumn(name = "sh_en_id")
    private List<StatusHistory> statusHistory = new ArrayList<>();

    @OneToMany(cascade = CascadeType.ALL)
    @JoinColumn(name = "cc_en_id")
    private List<CodeableConcept> type  = new ArrayList<>();

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "en_subject_re_id", referencedColumnName = "id")
    private Reference subject;

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "en_episodeofcare_re_id", referencedColumnName = "id")
    private Reference episodeOfCare;

    @OneToMany(cascade = CascadeType.ALL)
    @JoinColumn(name = "re_en_id")
    private List<Reference> appointment = new ArrayList<>();

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "en_pe_id")
    private Period period;

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "en_reasonreference_re_id", referencedColumnName = "id")
    private Reference reasonReference;

    @OneToMany(cascade = CascadeType.ALL)
    @JoinColumn(name = "di_en_id")
    private List<Diagnosis> diagnosis  = new ArrayList<>();

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "en_partof_re_id", referencedColumnName = "id")
    private Reference partOf;


    @OneToMany(cascade = CascadeType.ALL)
    @JoinColumn(name = "pp_en_id")
    private List<Participant> participant  = new ArrayList<>();
}

File: fhirstarter/spengerclient/src/app/models/Encounter.ts
export class Encounter {
  constructor(
    public id: string = "",
    public status: EncounterStatus = EncounterStatus.Planned,
    public period: Period = new Period(new Date(), new Date()),
    public type: Array<CodeableConcept> = [],
    public diagnosis: Array<Diagnosis> = [],
    public participant: Array<Participant> = [],
    public identifier: Array<Identifier> = [],
    public appointment: Array<Reference> = [],
    public subject: Reference = new Reference(),
    public episodeOfCare: Reference = new Reference(),
    public reasonReference: Reference = new Reference(),
    public partOf: Reference = new Reference(),
    /*public statusHistory: Array<StatusHistory> = []*/
  ) { }
}

/* export class StatusHistory {
  constructor(
    public period: Period = new Period(new Date("1999-01-01T00:00:00"), new Date("1999-01-01T00:00:00"),
    )
  ) { }
} */

  export class Period {
    constructor(public start: Date | null = null, public end: Date | null = null) {
      console.log("Periodclass: ", start, end);
    }
  }

export class CodeableConcept {
  constructor(public coding: Array<Coding> = [], public text: string = "") { }
}

export class Coding {
  constructor(
    public system: string = "",
    public version: string = "",
    public code: string = "",
    public display: string = "",
    public userSelected: boolean = false
  ) { }
}

export class Diagnosis {
  constructor(
    public condition: Reference = new Reference(),
    public use: CodeableConcept = new CodeableConcept(),
    public rank: number = 0
  ) { }
}

export class Participant {
  constructor(
    public type: Array<CodeableConcept> = [],
    public period: Period = new Period(new Date(), new Date()),
    public individual: Reference = new Reference()
  ) { }
}

export class Reference {  
  constructor(
    public reference: string = "",
    public type: string = "",
    public identifier: Identifier = new Identifier(),
    public display: string = ""
  ) { }
}

export class Identifier {
  constructor(public value: string = "") { }
}

export enum EncounterStatus {
  Planned = "planned",
  Arrived = "arrived",
  Triaged = "triaged",
  InProgress = "in-progress",
  OnLeave = "onleave",
  Finished = "finished",
  Cancelled = "cancelled",
}
